### PR TITLE
TRUNK-5362: Issue with method in EncounterService causing extreme slowness

### PR DESCRIPTION
**Pull request by kdaud**
## Description of what I changed
I have changed from vector class to using  ArrayList while implementing the search indexing criteria for patient identifier.

The search indexing criteria for 2.0.x version in the method `getEncountersByPatientIdentifier()` uses vector while searching for the patient identifier. Though the vector class maintains insertion order for the data but it is synchronized thus giving poor performance in searching, adding, delete and update of its elements.

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-5362

## Checklist: I completed these to help reviewers :)

- [ x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java [hunk 1]
```diff
@@ -281,19 +281,20 @@ public Encounter getEncounter(Integer encounterId) throws APIException {
...
 	 * @see org.openmrs.api.EncounterService#getEncountersByPatientIdentifier(java.lang.String)
 	 */
+	@Override
 	@Transactional(readOnly = true)
 	public List<Encounter> getEncountersByPatientIdentifier(String identifier) throws APIException {
 		if (identifier == null) {
 			throw new IllegalArgumentException("The 'identifier' parameter is required and cannot be null");
 		}
 		
-		List<Encounter> encs = new Vector<Encounter>();
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:290]: What do you base on to tell that this is the cause of the slowness?

**Comment by dkayiwa** [comment]
Is there a reason why you are not using the master branch of openmrs-core?

File: api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java (hunk 1 above)
Comment (kdaud) [reply api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:290]: The way how the search logic is implemented, the method searches actually  more criteria other than just patient identifier, which is not what the name of the method should be doing. And at the same time though the vector class maintains insertion order for the data but it is synchronized thus giving poor performance in searching, adding, delete and update of its elements.

**Comment by kdaud** [comment]
The issue is reported after an upgrade to 2.0.x so that is why have redirected the fix to that branch other than master.

File: api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java (hunk 1 above)
Comment (dkayiwa) [reply api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:290]: Did you in any way test this to confirm that it was slow before your changes but faster afterwards?

**Comment by dkayiwa** [comment]
Is it faster on master?

**Comment by sjmckee** [comment]
I don't believe just changing a Vector to an ArrayList fixes this issue.  In pre-2.0 versions of OpenMRS, this method ONLY searched by patient identifiers.  In post-2.0 versions, it searches across all aspects of the patient (patient identifiers, person names, person attributes), and this is what makes it extremely slow.  The method should only search based upon what the method name states...by identifier.

**Comment by kdaud** [comment]
Master actually is using ArrayList but as mentioned above, there is less significance.

I notice  things get slow when during the search indexing criteria . It looks like 6 queries are executed every time HibernatePatientDAO.getPatients(String query, boolean includeVoided, Integer start, Integer length) is executed, which is what gets called by EncounterServiceImpl.getEncountersByPatientIdentifier(). 

Now since “getEncountersByIPatientIdentifier” does more than searching by identifier my work around is to deprecate this method and then refactor it to have the ability to search only by identifier.

File: api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java (hunk 1 above)
Comment (kdaud) [reply api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:290]: I tried running against a deidentified copy of my database using a profiler but noticed a less significance.

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:290]: Then what you have done does not fix the reported problem.

**Pull request closed without merging**